plutus-core/plutus-ir/src/PlutusIR/Transform/Inline/Inline.hs

```diff
@@ -212,17 +214,114 @@ processTerm = handleTerm <=< traverseOf termSubtypes applyTypeSubstitution where
             -- actually have got rid of all of them!
             pure $ mkLet ann NonRec bs' t'
         -- This includes recursive let terms, we don't even consider inlining them at the moment
-        t ->
-            -- process all subterms first, so that the rhs won't be processed more than once. This
-            -- is important because otherwise the number of times we process them can grow
-            -- exponentially in the case that it has nested `let`s.
-            --
-            -- Then, consider call site inlining for each node that have gone through unconditional
-            -- inlining. Because `inlineSaturatedApp` traverses *all* application nodes for each
-            -- subterm, the runtime is quadratic for terms with a long chain of applications.
-            -- If we use the context-based approach like in GHC, this won't be a problem, so we may
-            -- consider that in the future.
-            inlineSaturatedApp =<< forMOf termSubterms t processTerm
+        t -> do
+            -- See note [Processing order of call site inlining]
+            let (hd, args) = splitApplication t
+                -- process the term args, ignore the type arguments
+                processArgs :: AppContext tyname name uni fun ann
+                    -> InlineM tyname name uni fun ann (AppContext tyname name uni fun ann)
+                processArgs (TermAppContext arg ann ctx) = do
+                    processedArg <- processTerm arg
+                    processedArgs <- processArgs ctx
+                    pure $ TermAppContext processedArg ann processedArgs
+                processArgs (TypeAppContext ty ann ctx) = do
+                    processedArgs <- processArgs ctx
+                    pure $ TypeAppContext ty ann processedArgs
+                processArgs AppContextEnd = pure AppContextEnd
+
+            case hd of
+                Var _ann name -> do
+                    gets (lookupVarInfo name) >>= \case
+                        Just varInfo -> do
+                            -- process the args if it's a variable that's in the map
+                            processedArgs <- processArgs args
+                            -- consider call site inlining since it's a variable.
+                            callSiteInline hd varInfo processedArgs
```

Comment:
I don't think this is correct, and this may be the reason for the free variable error. In addition to processing each arg, you should also need to process `hd`. Otherwise `processTerm` may never be applied to `hd`.

Reply:
Yes, this is definitely right. In particular, if we have `f a b` and `f` is being unconditionally inlined, then we've already removed the binding for `f` so we might end up with a free variable if we don't inline it.

You should just process all the parts of the application first, before doing anything else.

Reply:
I was thinking along that line as well about `hd`. Still working on it. Lots of weird things happening still.

Reply:
Still tempted to do the logging ticket first to be honest!

Reply:
Yeah, this is getting ridiculous. I really think we should do the logging thing. We are just constantly failing to work out what's going on and it's sucking up all the time.